session/schema_amender.go

```diff
@@ -0,0 +1,461 @@
+	}
+	if keyOnly {
+		return newIdxKey, []byte{}, nil
+	}
+
+	// Generate index value buf.
+	var containsNonBinaryString bool
+	for _, idxCol := range a.indexInfoAtCommit.Meta().Columns {
+		col := a.tblInfoAtCommit.Meta().Columns[idxCol.Offset]
+		if col.EvalType() == types.ETString && !mysql.HasBinaryFlag(col.Flag) {
+			containsNonBinaryString = true
+			break
+		}
+	}
+	newIdxVal, err := tablecodec.GenIndexValue(sctx.GetSessionVars().StmtCtx, a.tblInfoAtCommit.Meta(),
+		a.indexInfoAtCommit.Meta(), containsNonBinaryString, distinct, false, idxVals, kvHandle)
+	if err != nil {
+		logutil.Logger(ctx).Warn("amend generate index values failed", zap.Error(err))
+		return nil, nil, errors.Trace(err)
+	}
+	return newIdxKey, newIdxVal, nil
+}
+
+func (a *amendOperationAddIndex) processKey(ctx context.Context, sctx sessionctx.Context, keyOp pb.Op, key []byte,
+	kvMaps map[string][]byte) (resAdd tikv.CommitterMutations, err error) {
+	if a.AmendOpType == AmendNone || tablecodec.IsIndexKey(key) {
+		return
+	}
+	kvHandle, err := tablecodec.DecodeRowKey(key)
+	if err != nil {
+		logutil.Logger(ctx).Error("decode key error", zap.String("key", hex.EncodeToString(key)), zap.Error(err))
+		return resAdd, errors.Trace(err)
+	}
+
+	// Generated delete index key value.
+	if (a.AmendOpType == AmendNeedAddDelete || a.AmendOpType == AmendNeedAddDeleteAndInsert) && isDeleteOp(keyOp) {
+		newIdxKey, emptyVal, err := a.genIndexKeyValue(ctx, sctx, kvMaps, key, kvHandle, true)
+		if err != nil {
+			return resAdd, errors.Trace(err)
+		}
+		resAdd.Push(keyOp, newIdxKey, emptyVal, false)
+	}
+
+	// Generated insert index key value.
+	if (a.AmendOpType == AmendNeedAddDeleteAndInsert || a.AmendOpType == AmendNeedAddInsert) && isInsertOp(keyOp) {
+		newIdxKey, newIdxValue, err := a.genIndexKeyValue(ctx, sctx, kvMaps, key, kvHandle, false)
+		if err != nil {
+			return resAdd, errors.Trace(err)
+		}
+		// Check if the generated index keys are unique for unique index.
+		if a.indexInfoAtCommit.Meta().Unique {
+			if _, ok := a.processedNewIndexKeys[string(newIdxKey)]; ok {
+				return resAdd, errors.Trace(errors.Errorf("amend process key same key found for unique index=%v", a.indexInfoAtCommit.Meta().Name))
+			}
+		}
+		a.processedNewIndexKeys[string(newIdxKey)] = nil
+		resAdd.Push(keyOp, newIdxKey, newIdxValue, false)
+	}
+	return
+}
+
+// SchemaAmender is used to amend pessimistic transactions for schema change.
+type SchemaAmender struct {
+	sess *session
+}
+
+// NewSchemaAmenderForTikvTxn creates a schema amender for tikvTxn type.
+func NewSchemaAmenderForTikvTxn(sess *session) *SchemaAmender {
+	amender := &SchemaAmender{sess: sess}
+	return amender
+}
+
+func (s *SchemaAmender) getAmendableKeys(commitMutations tikv.CommitterMutations) []kv.Key {
+	kvKeys := make([]kv.Key, 0, len(commitMutations.GetKeys()))
+	for _, byteKey := range commitMutations.GetKeys() {
+		if tablecodec.IsIndexKey(byteKey) {
+			continue
+		}
+		kvKeys = append(kvKeys, byteKey)
```

Comment:
We can check the table id of the key, only get the keys for tables have AddIndex operation.